feat(Redirect): customRoutes now accept <Redirect />

kopax opened the pull request:
Fix [#2767](https://github.com/marmelab/react-admin/issues/2767)

kopax commented:
I am not familiar with the testing of the router, but testing the redirect work is like testing the `<Redirect />` itself.

Perhaps check if the `<Resource />` is renderer is easier and does the same.

fzaninotto commented:
Can't we *clone* the routes given by the developer instead of creating new ones and passing the developer's props to them?

kopax commented:
Do you mean this: 

```js
customRoutes
                        .filter(route => route.props.noLayout)
                        .map((route, index) => route.props.path ? (
                            <Route
                                key={index}
                                exact={route.props.exact}
                                path={route.props.path}
                                render={props =>
                                    this.renderCustomRoutesWithoutLayout(
                                        route,
                                        props
                                    )
                                }
                            />
                        ) :  cloneElement(route, route.props))
```
I was asking myself the same question for the `<Route />` implementation the line before.

What makes you prefer one on another (for each case)?

fzaninotto commented:
No, I mean don't even bother to test for the `path` prop, clone the `route` directly. Whether it's a `Route` or a `Redirect`, it should work.

kopax commented:
Sure, that's sounds better, I'll update.

kopax commented:
`CoreAdminRoute` ([instead of](https://github.com/marmelab/react-admin/blob/v2.6.0/packages/ra-core/src/CoreAdminRouter.js#L136)):

```js
{customRoutes
                        .filter(route => route.props.noLayout)
                        .map((route) => cloneElement(route, {
                          ...route.props ,
                          exact: props.path ? true : props.exact,
                          render: props =>
                            this.renderCustomRoutesWithoutLayout(
                              route,
                              props
                            )
                        }))}
```

`RoutesWithLayout` ([instead of](https://github.com/marmelab/react-admin/blob/v2.6.0/packages/ra-core/src/RoutesWithLayout.js#L20)):

```js
customRoutes &&
                customRoutes.map((route) => cloneElement(route, route.props))
```

- For `CoreAdminRoute`, is it voluntary to force the user to use `exact=true`, should it also concern the redirect?
- For `RoutesWithLayout`, why not just appending `customRoutes` instead of cloning ?

fzaninotto commented on packages/ra-core/src/RoutesWithLayout.js:
```diff
@@ -16,17 +16,7 @@ const RoutesWithLayout = ({
-                        children={route.props.children} // eslint-disable-line react/no-children-prop
-                    />
-                ))}
+            {customRoutes}
```
That won't work, `customRoutes` is an array.

fzaninotto commented on packages/ra-core/src/CoreAdminRouter.js:
```diff
@@ -1,7 +1,7 @@
 import PropTypes from 'prop-types';
 import { connect } from 'react-redux';
-import { Route, Switch } from 'react-router-dom';
+import { Route, Switch, Redirect } from 'react-router-dom';
```
not used anymore

fzaninotto commented on packages/ra-core/src/CoreAdminRouter.js:
```diff
@@ -135,19 +135,15 @@ export class CoreAdminRouter extends Component {
-                            />
-                        ))}
+                        .map((route) => cloneElement(route, {
+                          ...route.props ,
```
no need to pass the original route props

fzaninotto commented on packages/ra-core/src/RoutesWithLayout.js:
```diff
@@ -16,17 +16,7 @@ const RoutesWithLayout = ({
-                        children={route.props.children} // eslint-disable-line react/no-children-prop
-                    />
-                ))}
+            {customRoutes && customRoutes.map((route) => cloneElement(route, route.props))}
```
no need to pas the original route props

djhi commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
```
Please remove those `.idea` files

djhi commented on package-lock.json:
```diff
@@ -0,0 +1,8952 @@
+{
```
We use `yarn`. Please remove this file too

kopax replied:
Why not add a `.npmrc` with `package-lock=false` if you don't want this lock file?

djhi replied:
As said in the previous comment, we use `yarn`

kopax commented:
Sorry, I did it again. I had an issue when doing it the first time:

```bash
dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: git rm -r --cached .idea
rm '.idea/inspectionProfiles/Project_Default.xml'
rm '.idea/misc.xml'
rm '.idea/modules.xml'
rm '.idea/react-admin.iml'
rm '.idea/vcs.xml'
rm '.idea/workspace.xml'
dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: git rm -r --cached package-lock.json 
rm 'package-lock.json'
dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: git status
On branch next
Your branch is up to date with 'origin/next'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    .idea/inspectionProfiles/Project_Default.xml
	deleted:    .idea/misc.xml
	deleted:    .idea/modules.xml
	deleted:    .idea/react-admin.iml
	deleted:    .idea/vcs.xml
	deleted:    .idea/workspace.xml
	deleted:    package-lock.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.idea/
	package-lock.json

dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: a
dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: c 'chore(ignore): removed .idea and package-lock.json'
On branch next
Your branch is up to date with 'origin/next'.

nothing to commit, working tree clean
dka@dev-12:[~/workspace/github.com/kopax/react-admin (next)]: p next
Everything up-to-date
```

I had to really `rm -rf .idea package-lock.json` and then do that again, not sure why it failed. Would it be possible to run and add in this commit:

1.  `echo /.idea >> .gitignore` 
1. `echo package-lock=false > .npmrc`

- *(1)* will allow to use `.idea` without worrying of pushing on the remote the `.idea`.
- *(2)* will allow prevent generating the `package-lock.json` to everyone. It is different than adding to `.gitignore` the `package-lock.json`, because no one will be able to have a lock file.

There's two option, either **react-admin** is a framework, or it is an application, if it's a framework, it make sens to not have a lock file as this is for application.

fzaninotto commented:
1. No, we won't search all possible IDEs and add all the files that they create in a project into our `.gitignore`. It's your IDE, it's your job to manage its files. IntelliJ is no exception.
2. Sorry, we won't do that either. If you want to contribute to react-admin, please use the tools we use, or assume the consequences. Once again, we can't support every tool chain in the world, so we chose yarn. I don't want to learn about package.lock and npmrc, I prefer to spend that time actually improving react-admin.

React-admin is a library, but that's not what motivates us to add a lock file. It's the fact that dependency management is permissive in JS, and a lock file is the only guarantee we have that something that worked yesterday still works today. 

Also, this discussion has nothing to see with your original PR, so if you want to continue it, I suggest you open a new RFC issue.

fzaninotto commented on packages/ra-core/src/CoreAdminRouter.js:
```diff
@@ -135,19 +135,14 @@ export class CoreAdminRouter extends Component {
-                            />
-                        ))}
+                        .map((route) => cloneElement(route, {
+                          exact: true,
```
why did you override the route's `exact` prop? That's a regression.

kopax replied:
I though it was written before `<Route exact ... />` but yes after verifying the history it is a mistake again.

fzaninotto commented on packages/ra-core/src/RoutesWithLayout.js:
```diff
@@ -16,17 +16,7 @@ const RoutesWithLayout = ({
-                        children={route.props.children} // eslint-disable-line react/no-children-prop
-                    />
-                ))}
+            {customRoutes && customRoutes.map((route) => cloneElement(route))}
```
Are you sure that React won't complain about missing `key` here? Did you test that code?

kopax replied:
Yes I did, it does not through any warning.

fzaninotto replied:
Well, I should have checked before merging: your code does throw a warning in the simple demo:

```
warning.js:33 Warning: Each child in an array or iterator should have a unique "key" prop.

Check the render method of `CoreAdminRouter`. See https://fb.me/react-warning-keys for more information.
    in Route
    in CoreAdminRouter (created by Connect(CoreAdminRouter))
    in Connect(CoreAdminRouter) (created by getContext(Connect(CoreAdminRouter)))
    in getContext(Connect(CoreAdminRouter)) (created by Route)
    in Route (created by CoreAdmin)
    in Switch (created by CoreAdmin)
    in Router (created by ConnectedRouter)
    in ConnectedRouter (created by CoreAdmin)
    in TranslationProviderView (created by Connect(TranslationProviderView))
    in Connect(TranslationProviderView) (created by CoreAdmin)
    in Provider (created by CoreAdmin)
    in CoreAdmin (created by withContext(CoreAdmin))
    in withContext(CoreAdmin)
```

I'll have to fix that in next.

fzaninotto commented:
Thanks!

kopax commented:
@fzaninotto thanks to you too for accepting this feature.
